src/node_api_jsrt.cc

```diff
@@ -221,43 +265,100 @@ namespace jsrtimpl {
 
-#define RETURN_STATUS_IF_FALSE(condition, status)                       \
-  do {                                                                  \
-    if (!(condition)) {                                                 \
-      return napi_set_last_error((status));                             \
-    }                                                                   \
-  } while(0)
+    if (keyType == JsString) {
+      size_t length;
+      CHECK_JSRT_EXPECTED(
+        JsCopyStringUtf8(key, nullptr, 0, &length), napi_string_expected);
 
-#define CHECK_ARG(arg)                                                  \
-  RETURN_STATUS_IF_FALSE((arg), napi_invalid_arg)
+      std::vector<uint8_t> name;
+      name.reserve(length + 1);
+      CHECK_JSRT(JsCopyStringUtf8(key, name.data(), length + 1, &length));
 
-#define CHECK_JSRT(expr)                                                \
-  do {                                                                  \
-    JsErrorCode err = (expr);                                           \
-    if (err != JsNoError) return napi_set_last_error(err);              \
-  } while(0)
+      CHECK_JSRT(JsCreatePropertyIdUtf8(
+        reinterpret_cast<char*>(name.data()), length, propertyId));
+    } else {
+      CHECK_JSRT(JsGetPropertyIdFromSymbol(key, propertyId));
+    }
+    return napi_ok;
+  }
 
-#define CHECK_JSRT_EXPECTED(expr, expected)                             \
-  do {                                                                  \
-    JsErrorCode err = (expr);                                           \
-    if (err == JsErrorInvalidArgument)                                  \
-      return napi_set_last_error(expected);                             \
-    if (err != JsNoError) return napi_set_last_error(err);              \
-  } while(0)
+  inline napi_status
+  JsPropertyIdFromPropertyDescriptor(const napi_property_descriptor* p,
+                                     JsPropertyIdRef* propertyId) {
+    if (p->utf8name != nullptr) {
+      CHECK_JSRT(JsCreatePropertyIdUtf8(
+        p->utf8name, strlen(p->utf8name), propertyId));
+      return napi_ok;
+    } else {
+      return JsPropertyIdFromKey(p->name, propertyId);
+    }
+  }
 
-// This does not call napi_set_last_error because the expression
-// is assumed to be a NAPI function call that already did.
-#define CHECK_NAPI(expr)                                                \
-  do {                                                                  \
-    napi_status status = (expr);                                        \
-    if (status != napi_ok) return status;                               \
-  } while(0)
+  inline napi_status
+  JsNameValueFromPropertyDescriptor(const napi_property_descriptor* p,
+                                    napi_value* name) {
+    if (p->utf8name != nullptr) {
+      CHECK_JSRT(JsCreateStringUtf8(
+        reinterpret_cast<const uint8_t*>(p->utf8name),
+        strlen(p->utf8name),
```

Comment:
If you want to use something other than a null-terminated UTF-8 string as the property key, you have the option to create a `napi_value` first and assign that to the `napi_property_descriptor.name` field instead of using the `utf8name` field.